apps/vault.go

```diff
@@ -1,10 +1,251 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package apps
 
 import (
+	"fmt"
+
 	"github.com/GoogleCloudPlatform/ops-agent/confgenerator"
 	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/fluentbit"
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/otel"
 )
 
+type MetricsReceiverVault struct {
+	confgenerator.ConfigComponent          `yaml:",inline"`
+	confgenerator.MetricsReceiverShared    `yaml:",inline"`
+	confgenerator.MetricsReceiverSharedTLS `yaml:",inline"`
+
+	Token       string `yaml:"token"`
+	Endpoint    string `yaml:"endpoint" validate:"omitempty,hostname_port"`
+	MetricsPath string `yaml:"metrics_path" validate:"omitempty,startswith=/"`
+}
+
+const (
+	defaultVaultEndpoint = "localhost:8200"
+	defaultMetricsPath   = "/v1/sys/metrics"
+)
+
+func (r MetricsReceiverVault) Type() string {
+	return "vault"
+}
+
+func (r MetricsReceiverVault) Pipelines() []otel.Pipeline {
+	if r.Endpoint == "" {
+		r.Endpoint = defaultVaultEndpoint
+	}
+	if r.MetricsPath == "" {
+		r.MetricsPath = defaultMetricsPath
+	}
+
+	tlsConfig := r.TLSConfig(true)
+
+	scrapeConfig := map[string]interface{}{
+		"job_name":        "vault",
+		"scrape_interval": r.CollectionIntervalString(),
+		"metrics_path":    r.MetricsPath,
+		"static_configs": []map[string]interface{}{{
+			"targets": []string{r.Endpoint},
+		}},
+	}
+
+	if r.Token != "" {
+		scrapeConfig["scheme"] = "https"
+		scrapeConfig["authorization"] = map[string]interface{}{
+			"credentials": r.Token,
+			"type":        "Bearer",
+		}
+	}
+	if tlsConfig["insecure"] == false {
+		scrapeConfig["scheme"] = "https"
+	}
+	delete(tlsConfig, "insecure")
+	scrapeConfig["tls_config"] = tlsConfig
+
+	includeMetrics := []string{}
+
+	storageMetricTransforms, newStorageMetricNames := r.addStorageMetrics()
+	metricDetailTransforms, newMetricNames := r.getMetricTransforms()
+
+	includeMetrics = append(includeMetrics, newStorageMetricNames...)
+	includeMetrics = append(includeMetrics, newMetricNames...)
+
+	return []otel.Pipeline{{
+		Receiver: otel.Component{
+			Type: "prometheus",
+			Config: map[string]interface{}{
+				"config": map[string]interface{}{
+					"scrape_configs": []map[string]interface{}{
+						scrapeConfig,
+					},
+				},
+			},
+		},
+		Processors: []otel.Component{
+			storageMetricTransforms,
```

Comment:
Unless there's a reason I'm missing, should refactor to only create one Transform processor for the two different sets of actions

Reply:
Updated